test/test_images/sender_gcpbroker/main.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2019 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go"
+	transport "github.com/cloudevents/sdk-go/pkg/cloudevents/transport/http"
+	"go.opencensus.io/trace"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+
+	"github.com/google/knative-gcp/pkg/kncloudevents"
+	"github.com/google/knative-gcp/test/e2e/lib"
+)
+
+const (
+	brokerURLEnvVar = "BROKER_URL"
+)
+
+// defaultRetry represents that there will be 3 iterations.
+// The duration starts from 30s and is multiplied by factor 1.0 for each iteration.
+var defaultRetry = wait.Backoff{
+	Steps:    3,
+	Duration: 30 * time.Second,
+	Factor:   1.0,
+	// The sleep at each iteration is the duration plus an additional
+	// amount chosen uniformly at random from the interval between 0 and jitter*duration.
+	Jitter: 2.0,
+}
+
+func main() {
```

Comment:
Is it possible to add this retry functionality to the existing sender? I think it would allow easier reuse.

Reply:
Make senses. I'll change it. Maybe pass a new env in the pod, indicating whether you want to use the try mode for the sender pod or not?

Reply:
Yes. Thanks!

Reply:
I'm not sure that we have any tests where we would currently want to disable retries.

Reply:
Maybe for metrics test, we don't want to enable retries? otherwise the metric count might be different from what we expected.

Reply:
No, it should be fine. If a failure results in additional metrics it will fail the test either way.

Reply:
Checked the usage, sender Pod is only used in three tests (I thought it will be used in more places...) As all three tests can use retry mode, I directly add the them without env.

Reply:
The sender will be used for both gcp broker and pubsub channel based broker, right? Do we need the retry mode for the pubsub channel based broker?

Reply:
```
func SenderJob(name string, envVars []corev1.EnvVar) *batchv1.Job {
	return baseJob(name, "sender", envVars)
}
```
is only used in three broker test. Let me double check for it.

Reply:
Yep. That's used in broker but that lib method is for both gcp broker and pubsub channel based broker(just currently pubsub channel based broker e2e tests get skipped because of issues in eventing)

Reply:
Ah, we also use it in BrokerWithPubSubChannel test (so, three gcp broker tests + one pubsub channel test). I think it would be fine to also retry in pubsub channel (as we indicate only retry for specific error, so it most likely won't qualify for retry).

Reply:
Retry is just a work-around for gcp broker e2e tests. Usually broker with pubsub channel should not use it and if it start to use it, it means something goes wrong and devs should investigate what happens. I'd prefer not enabling retryable sender for broker with pubsub channel.

Reply:
Sounds good. I am find to change it.